include/RAJA/policy/hip/policy.hpp

```diff
@@ -396,21 +333,190 @@ struct HipDimHelper<2>{
+  RAJA_DEVICE
+  static inline constexpr
+  auto size() { return static_cast<IdxT>(t_block_size); }
+};
+/// unless t_block_size is 0 then block_size is dynamic
+template<int dim>
+struct HipIndexThread<dim, 0> {
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  inline static constexpr
+  auto index() { return static_cast<IdxT>(HipDimHelper<dim>::get(threadIdx)); }
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  inline static constexpr
+  auto size() { return static_cast<IdxT>(HipDimHelper<dim>::get(blockDim)); }
+};
+
+/// Type representing block indexing within a grid
+/// grid_size is fixed
+template<int dim, HipDimIdxT t_grid_size>
+struct HipIndexBlock {
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  inline static constexpr
+  auto index() { return static_cast<IdxT>(HipDimHelper<dim>::get(blockIdx)); }
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  inline static constexpr
+  auto size() { return static_cast<IdxT>(t_grid_size); }
+};
+/// unless t_grid_size is 0 then grid_size is dynamic
+template<int dim>
+struct HipIndexBlock<dim, 0> {
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  inline static constexpr
+  auto index() { return static_cast<IdxT>(HipDimHelper<dim>::get(blockIdx)); }
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_DEVICE
+  inline static constexpr
+  auto size() { return static_cast<IdxT>(HipDimHelper<dim>::get(gridDim)); }
+};
+
+/// Type representing thread indexing within a grid
+template<int dim, HipDimIdxT t_block_size, HipDimIdxT t_grid_size>
+struct HipIndexGlobal {
+
+  using IndexThread = HipIndexThread<dim, t_block_size>;
+  using IndexBlock = HipIndexBlock<dim, t_block_size>;
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_HOST_DEVICE
+  inline static constexpr
+  auto index() { return static_cast<IdxT>(IndexThread::template index<IdxT>()) +
+                        static_cast<IdxT>(IndexThread::template size<IdxT>()) *
+                        static_cast<IdxT>(IndexBlock::template index<IdxT>()) ; }
+
+  template < typename IdxT = HipDimIdxT >
+  RAJA_HOST_DEVICE
+  inline static constexpr
+  auto size() { return static_cast<IdxT>(IndexThread::template size<IdxT>()) *
+                       static_cast<IdxT>(IndexBlock::template size<IdxT>()) ; }
+};
+
+
+template<typename Indexer>
+struct HipIndexDirect {
+  using indexer_type = Indexer;
+};
+
+template<typename Indexer>
+struct HipIndexLoop {
+  using indexer_type = Indexer;
+};
+
 } // namespace internal
 
 
+/*!
+ * Maps segment indices to HIP threads.
+ * This is the lowest overhead mapping, but requires that there are enough
+ * physical threads to fit all of the direct map requests.
+ * For example, a segment of size 2000 will not fit, and trigger a runtime
+ * error.
+ */
+template<int ... dim>
+struct hip_thread_xyz_direct{};
+
+using hip_thread_x_direct = internal::HipIndexDirect<internal::HipIndexThread<0, 0>>; // hip_thread_xyz_direct<0>;
```

Comment:
I would like to add the ability to set the size template param, but I don't want to break existing aliases. I will add another set of alias templates to handle this.